File: web/js/sim/lidar.js
// LiDAR: N rays over a forward FOV arc against wall segments, accelerated by
// a uniform spatial grid + DDA walk. MUST match racing/sim/lidar.py.

const EPS = 1e-12;

// Ray (ox,oz,dx,dz) vs segment (ax,az)-(bx,bz). Returns t along ray or -1.
export function raySegment(ox, oz, dx, dz, ax, az, bx, bz) {
  const ex = bx - ax;
  const ez = bz - az;
  const denom = dx * ez - dz * ex;
  if (denom > -EPS && denom < EPS) {
    return -1.0;
  }
  const wx = ax - ox;
  const wz = az - oz;
  const t = (wx * ez - wz * ex) / denom;
  const u = (wx * dz - wz * dx) / denom;
  if (t >= 0.0 && u >= 0.0 && u <= 1.0) {
    return t;
  }
  return -1.0;
}

export class Lidar {
  constructor(track, C) {
    this.C = C;
    this.cell = C.LIDAR_GRID_CELL;
    // Segment soup from both walls (closed loops).
    const segs = [];
    for (const wall of [track.left_wall, track.right_wall]) {
      const n = wall.length;
      for (let i = 0; i < n; i++) {
        const a = wall[i];
        const b = wall[(i + 1) % n];
        segs.push(a[0], a[1], b[0], b[1]);
      }
    }
    this.segs = new Float64Array(segs);
    this.numSegs = segs.length / 4;
    // Grid: map "ix,iz" -> array of segment indices (bbox rasterization).
    this.grid = new Map();
    for (let i = 0; i < this.numSegs; i++) {
      const o = i * 4;
      const x0 = Math.min(this.segs[o], this.segs[o + 2]);
      const x1 = Math.max(this.segs[o], this.segs[o + 2]);
      const z0 = Math.min(this.segs[o + 1], this.segs[o + 3]);
      const z1 = Math.max(this.segs[o + 1], this.segs[o + 3]);
      const ix0 = Math.floor(x0 / this.cell);
      const ix1 = Math.floor(x1 / this.cell);
      const iz0 = Math.floor(z0 / this.cell);
      const iz1 = Math.floor(z1 / this.cell);
      for (let ix = ix0; ix <= ix1; ix++) {
        for (let iz = iz0; iz <= iz1; iz++) {
          const key = `${ix},${iz}`;
          let list = this.grid.get(key);
          if (!list) {
            list = [];
            this.grid.set(key, list);
          }
          list.push(i);
        }
      }
    }
    this.stamp = new Int32Array(this.numSegs).fill(-1);
    this.rayId = 0;
  }

  // Single ray distance, capped at LIDAR_MAX_RANGE. DDA over grid cells with
  // early exit once the best hit is closer than the next cell boundary.
  castRay(ox, oz, dx, dz) {
    const { cell, grid, segs, stamp, C } = this;
    const maxR = C.LIDAR_MAX_RANGE;
    const id = this.rayId++;
    let bestT = maxR;

    let ix = Math.floor(ox / cell);
    let iz = Math.floor(oz / cell);
    const stepX = dx > 0.0 ? 1 : -1;
    const stepZ = dz > 0.0 ? 1 : -1;
    const tDeltaX = dx !== 0.0 ? Math.abs(cell / dx) : Infinity;
    const tDeltaZ = dz !== 0.0 ? Math.abs(cell / dz) : Infinity;
    let tMaxX =
      dx !== 0.0
        ? ((dx > 0.0 ? (ix + 1) * cell - ox : ox - ix * cell) / Math.abs(dx))
        : Infinity;
    let tMaxZ =
      dz !== 0.0
        ? ((dz > 0.0 ? (iz + 1) * cell - oz : oz - iz * cell) / Math.abs(dz))
        : Infinity;

    let tEntry = 0.0;
    for (let guard = 0; guard < 64; guard++) {
      if (tEntry > bestT) {
        break;
      }
      const list = grid.get(`${ix},${iz}`);
      if (list) {
        for (let k = 0; k < list.length; k++) {
          const si = list[k];
          if (stamp[si] === id) {
            continue;
          }
          stamp[si] = id;
          const o = si * 4;
          const t = raySegment(
            ox, oz, dx, dz, segs[o], segs[o + 1], segs[o + 2], segs[o + 3]
          );
          if (t >= 0.0 && t < bestT) {
            bestT = t;
          }
        }
      }
      if (tMaxX < tMaxZ) {
        tEntry = tMaxX;
        tMaxX += tDeltaX;
        ix += stepX;
      } else {
        tEntry = tMaxZ;
        tMaxZ += tDeltaZ;
        iz += stepZ;
      }
      if (tEntry > maxR) {
        break;
      }
    }
    return bestT;
  }

  // Returns Float64Array of LIDAR_NUM_RAYS normalized [0,1] distances.
  // Ray i angle: heading - FOV/2 + FOV * i / (N - 1).
  scan(x, z, heading) {
    const { C } = this;
    const n = C.LIDAR_NUM_RAYS;
    const out = new Float64Array(n);
    for (let i = 0; i < n; i++) {
      const ang = heading - C.LIDAR_FOV / 2.0 + (C.LIDAR_FOV * i) / (n - 1);
      const d = this.castRay(x, z, Math.cos(ang), Math.sin(ang));
      out[i] = d / C.LIDAR_MAX_RANGE;
    }
    return out;
  }
}

File: racing/sim/lidar.py
"""LiDAR: N rays over a forward FOV arc against wall segments, accelerated by
a uniform spatial grid + DDA walk.

MUST match web/js/sim/lidar.js line by line (parity-tested).
"""

import math

from racing import constants as C
from racing.sim.collision import CAR_CIRCLE_OFFSETS, CAR_CIRCLE_RADIUS

EPS = 1e-12


def ray_circle(ox, oz, dx, dz, cx, cz, r):
    """Ray (unit dir) vs circle; nearest t >= 0 along the ray or -1.0.
    Returns 0.0 when the origin is inside the circle."""
    mx = ox - cx
    mz = oz - cz
    b = mx * dx + mz * dz
    c = mx * mx + mz * mz - r * r
    if c > 0.0 and b > 0.0:
        return -1.0
    disc = b * b - c
    if disc < 0.0:
        return -1.0
    t = -b - math.sqrt(disc)
    if t < 0.0:
        t = 0.0
    return t


def ray_segment(ox, oz, dx, dz, ax, az, bx, bz):
    """Ray vs segment; returns t along ray or -1.0."""
    ex = bx - ax
    ez = bz - az
    denom = dx * ez - dz * ex
    if -EPS < denom < EPS:
        return -1.0
    wx = ax - ox
    wz = az - oz
    t = (wx * ez - wz * ex) / denom
    u = (wx * dz - wz * dx) / denom
    if t >= 0.0 and 0.0 <= u <= 1.0:
        return t
    return -1.0


class Lidar:
    def __init__(self, track: dict):
        self.cell = C.LIDAR_GRID_CELL
        segs = []
        for wall in (track["left_wall"], track["right_wall"]):
            n = len(wall)
            for i in range(n):
                a = wall[i]
                b = wall[(i + 1) % n]
                segs.append((a[0], a[1], b[0], b[1]))
        self.segs = segs
        self.grid: dict[tuple[int, int], list[int]] = {}
        for i, (ax, az, bx, bz) in enumerate(segs):
            ix0 = math.floor(min(ax, bx) / self.cell)
            ix1 = math.floor(max(ax, bx) / self.cell)
            iz0 = math.floor(min(az, bz) / self.cell)
            iz1 = math.floor(max(az, bz) / self.cell)
            for ix in range(ix0, ix1 + 1):
                for iz in range(iz0, iz1 + 1):
                    self.grid.setdefault((ix, iz), []).append(i)
        self.stamp = [-1] * len(segs)
        self.ray_id = 0

    def cast_ray(self, ox, oz, dx, dz):
        """Single ray distance, capped at LIDAR_MAX_RANGE. DDA over grid cells
        with early exit once the best hit is closer than the next boundary."""
        cell = self.cell
        max_r = C.LIDAR_MAX_RANGE
        ray_id = self.ray_id
        self.ray_id += 1
        best_t = max_r

        ix = math.floor(ox / cell)
        iz = math.floor(oz / cell)
        step_x = 1 if dx > 0.0 else -1
        step_z = 1 if dz > 0.0 else -1
        t_delta_x = abs(cell / dx) if dx != 0.0 else math.inf
        t_delta_z = abs(cell / dz) if dz != 0.0 else math.inf
        if dx != 0.0:
            t_max_x = ((ix + 1) * cell - ox if dx > 0.0 else ox - ix * cell) / abs(dx)
        else:
            t_max_x = math.inf
        if dz != 0.0:
            t_max_z = ((iz + 1) * cell - oz if dz > 0.0 else oz - iz * cell) / abs(dz)
        else:
            t_max_z = math.inf

        t_entry = 0.0
        for _ in range(64):
            if t_entry > best_t:
                break
            lst = self.grid.get((ix, iz))
            if lst is not None:
                for si in lst:
                    if self.stamp[si] == ray_id:
                        continue
                    self.stamp[si] = ray_id
                    ax, az, bx, bz = self.segs[si]
                    t = ray_segment(ox, oz, dx, dz, ax, az, bx, bz)
                    if t >= 0.0 and t < best_t:
                        best_t = t
            if t_max_x < t_max_z:
                t_entry = t_max_x
                t_max_x += t_delta_x
                ix += step_x
            else:
                t_entry = t_max_z
                t_max_z += t_delta_z
                iz += step_z
            if t_entry > max_r:
                break
        return best_t

    def scan(self, x, z, heading, cars=None):
        """List of LIDAR_NUM_RAYS normalized [0,1] distances.
        Ray i angle: heading - FOV/2 + FOV * i / (N - 1).

        cars: optional iterable of opponent poses (objects with .x, .z,
        .heading); each is sensed as its collision capsule, so rays
        report the nearer of wall or car."""
        n = C.LIDAR_NUM_RAYS
        circles = []
        if cars:
            for car in cars:
                fx = math.cos(car.heading)
                fz = math.sin(car.heading)
                for o in CAR_CIRCLE_OFFSETS:
                    circles.append((car.x + fx * o, car.z + fz * o))
        out = []
        for i in range(n):
            ang = heading - C.LIDAR_FOV / 2.0 + (C.LIDAR_FOV * i) / (n - 1)
            dx = math.cos(ang)
            dz = math.sin(ang)
            d = self.cast_ray(x, z, dx, dz)
            for cx, cz in circles:
                t = ray_circle(x, z, dx, dz, cx, cz, CAR_CIRCLE_RADIUS)
                if 0.0 <= t < d:
                    d = t
            out.append(d / C.LIDAR_MAX_RANGE)
        return out
